src/core/core_tags.js

```diff
@@ -0,0 +1,410 @@
+            var index = { };
+            var items = [ ];
+            var i, n, t;
+            var item, tag, tags, tagsRest, list, missingIndex;
+
+            for(i = 0; i < args.length; i++) {
+                tag = args[i];
+
+                if (tag instanceof Array) {
+                    if (tag.length === 0)
+                        continue;
+
+                    if (tag.length === 1) {
+                        tag = tag[0];
+                    } else {
+                        // check if all indexes are in present
+                        missingIndex = false;
+                        for(t = 0; t < tag.length; t++) {
+                            if (! this._index[tag[t]]) {
+                                missingIndex = true;
+                                break;
+                            }
+                        }
+                        if (missingIndex)
+                            continue;
+
+                        // sort tags by least number of matches first
+                        tags = tag.slice(0).sort(function(a, b) {
+                            return self._index[a].list.length - self._index[b].list.length;
+                        });
+
+                        // remainder of tags for `has` checks
+                        tagsRest = tags.slice(1);
+                        if (tagsRest.length === 1)
+                            tagsRest = tagsRest[0];
+
+                        for(n = 0; n < this._index[tags[0]].list.length; n++) {
+                            item = this._index[tags[0]].list[n];
+                            if ((this._key ? ! index[item[this._key]] : (items.indexOf(item) === -1)) && item.tags.has(tagsRest)) {
+                                if (this._key)
+                                    index[item[this._key]] = true;
+                                items.push(item);
+                            }
+                        }
+
+                        continue;
+                    }
+                }
+
+                if (tag && typeof(tag) === 'string' && this._index[tag]) {
+                    for(n = 0; n < this._index[tag].list.length; n++) {
+                        item = this._index[tag].list[n];
+
+                        if (this._key) {
+                            if (! index[item[this._key]]) {
+                                index[item[this._key]] = true;
+                                items.push(item);
+                            }
+                        } else if (items.indexOf(item) === -1) {
+                            items.push(item);
+                        }
+                    }
+                }
+            }
+
+            return items;
+        }
+    };
+
+
+
+    /**
+    * @name pc.Tags
+    * @class Set of tag names
+    * @constructor Create an instance of an Tags.
+    * @param {Object} parent (optional) Parent object who tags belong to.
+    * Note: Tags are used as addition of `pc.Entity` and `pc.Asset` as `tags` field.
```

Comment:
TagsCache and Tags is implemented to be agnostic regarding what it's tagging. So can be added later, although currently it is not part of this PR, as there are few questions of where TagsCache should sit on: Application or Scene, or Root node?
As currently it is: `app.root.findByName`, so makes sense to do: `app.root.findByTag`, although root is a `GraphNode`, so a bit of technical design is required here.